src/rgw/rgw_rados.cc

```diff
@@ -3671,7 +3671,7 @@ int RGWRados::fetch_remote_obj(RGWObjectCtx& obj_ctx,
 
   if (copy_if_newer) {
     /* need to get mtime for destination */
-    ret = get_obj_state(&obj_ctx, dest_bucket_info, dest_obj, &dest_state, false, null_yield);
+    ret = get_obj_state(&obj_ctx, dest_bucket_info, src_obj, &dest_state, true, null_yield);
```

Comment:
ok i'm glad this fix works! before merge though, i'd like to find a way to preserve the ability for `fetch_remote_obj()` to use a different src_obj and dest_obj

my only idea here is to add two separate args to `fetch_remote_obj()` (which unfortunately already has a million arguments) for `stat_dest_obj` and `stat_follow_olh`

then for normal multisite replication, we'd pass `stat_dest_obj=dest_obj` and `stat_follow_olh=false`

and for the archive zone sync module, `RGWArchiveDataSyncModule::sync_object()` would pass `stat_dest_obj=key` and `stat_follow_olh=true` in the `/* force version if not set */` case

Reply:
@cbodley @smanjara I agree that we should make sure `fetch_remote_obj()` still does what it does now, but does this fix even work if there are multiple concurrent writes to the different zones on the same object? Seems to me that following the olh for mtime will only work consistently if there's only a single change. Maybe the way to go would be to identify the correct `dest` by searching for the `mtime` at the sync module level?

Reply:
i'm not sure what is meant by finding the correct change;  is it related to requiring archive of all unique versions of an object?  I'm unconvinced that this is a requirement--but if it is, I'm not certain that modification time is a valid selector for same

Reply:
@mattbenjamin the problem here is archiving too many versions of the same object instance. The fix here (if I understand it correctly) tries to eliminate it by identifying that that current versioned head object (on the archive zone) has the same mtime as the source object (which is not necessarily versioned). However, I don't believe that checking the current head object is reliable.

Reply:
> but does this fix even work if there are multiple concurrent writes to the different zones on the same object?

it doesn't, no. with concurrent writes to a non-versioned bucket on active-active zones A and B, replication will converge on the copy with the latest mtime. an archive zone C has no way to guarantee that it can fetch both copies of the object, because the replication between A and B is racing to overwrite one of them

even considering a simpler configuration with a single zone A and archive zone C, what guarantees can the archive zone make about its ability to archive overwrites to an object? if zone A overwrites the object every 10 minutes, and the archive zone's replication delta is lower than that, then it _should_ archive every copy. but the time between overwrites could be arbitratily small, and the replication delta could be arbitrarily large

> Maybe the way to go would be to identify the correct dest by searching for the mtime at the sync module level?

even if we do find a copy older than the 'current' archived version and include that mtime in the If-Modified-Since header, if the replication between A and B converges first, we'd still fetch two copies of that latest version and end up with duplicates

Reply:
I made essentially this argument in a conversation with Casey last week--with the quibble that it satisfies "works" for some definition of consistency.  It appeared to me that, as above, it feels too strong to require that every version that existed transiently needs to be captured at the archive zone (but someone might ask for it);  meanwhile, it feels almost expected that we would be able to remove duplicates, especially ones arriving at the same time, but maybe in general;  with some bound on the number of versions we permit in the archive zone, we could use etag/checksum checks to do that?